Shell/Builtin.cpp

```diff
@@ -0,0 +1,743 @@
+        } else {
+            StringBuilder builder;
+            builder.append(cwd);
+            builder.append('/');
+            builder.append(arg_path);
+            new_path = builder.to_string();
+        }
+    }
+
+    LexicalPath lexical_path(new_path);
+    if (!lexical_path.is_valid()) {
+        printf("LexicalPath failed to canonicalize '%s'\n", new_path.characters());
+        return 1;
+    }
+    const char* path = lexical_path.string().characters();
+
+    struct stat st;
+    int rc = stat(path, &st);
+    if (rc < 0) {
+        printf("stat(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    if (!S_ISDIR(st.st_mode)) {
+        printf("Not a directory: %s\n", path);
+        return 1;
+    }
+    rc = chdir(path);
+    if (rc < 0) {
+        printf("chdir(%s) failed: %s\n", path, strerror(errno));
+        return 1;
+    }
+    setenv("OLDPWD", cwd.characters(), 1);
+    cwd = lexical_path.string();
+    setenv("PWD", cwd.characters(), 1);
+    return 0;
+}
+
+int Shell::builtin_cdh(int argc, const char** argv)
+{
+    int index = -1;
+
+    Core::ArgsParser parser;
+    parser.add_positional_argument(index, "Index of the cd history entry (leave out for a list)", "index", Core::ArgsParser::Required::No);
+
+    if (!parser.parse(argc, const_cast<char**>(argv), false))
+        return 1;
+
+    if (index == -1) {
+        if (cd_history.size() == 0) {
+            printf("cdh: no history available\n");
+            return 0;
+        }
+
+        for (int i = cd_history.size() - 1; i >= 0; --i)
+            printf("%lu: %s\n", cd_history.size() - i, cd_history.at(i).characters());
+        return 0;
+    }
+
+    if (index < 1 || (size_t)index > cd_history.size()) {
+        fprintf(stderr, "cdh: history index out of bounds: %d not in (0, %zu)\n", index, cd_history.size());
+        return 1;
+    }
+
+    const char* path = cd_history.at(cd_history.size() - index).characters();
+    const char* cd_args[] = { "cd", path };
+    return Shell::builtin_cd(2, cd_args);
+}
+
+int Shell::builtin_dirs(int argc, const char** argv)
+{
+    // The first directory in the stack is ALWAYS the current directory
+    directory_stack.at(0) = cwd.characters();
+
+    if (argc == 1) {
+        for (auto& directory : directory_stack) {
+            print_path(directory);
+            fputc(' ', stdout);
+        }
+
+        printf("\n");
+        return 0;
+    }
+
+    bool clear = false;
+    bool print = false;
+    bool number_when_printing = false;
+
+    Vector<const char*> paths;
+
+    Core::ArgsParser parser;
+    parser.add_option(clear, "Clear the directory stack", "clear", 'c');
+    parser.add_option(print, "Print directory entries", "print", 'p');
+    parser.add_option(number_when_printing, "Number the directories in the stack when printing", "number", 'v');
+    parser.add_positional_argument(paths, "Extra paths to put on the stack", "paths", Core::ArgsParser::Required::No);
```

Comment:
```suggestion
    parser.add_positional_argument(paths, "Extra paths to put on the stack", "path", Core::ArgsParser::Required::No);
```

Singular, as discussed over in the other PR.